udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,123 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
+/**
+ * Returns a collection of TimeRange objects to indicate available meeting times
+ * for indicated people and duration given through a request object.
+ */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Create an output collection holding a starting full day block.
+    Collection<TimeRange> rangeOutput = new HashSet<>();
+    TimeRange fullDay = TimeRange.fromStartDuration(0, 24 * 60);
+    rangeOutput.add(fullDay);
+
+    // Obtain data from the request.
+    Collection<String> attendees = request.getAttendees();
+    long requestDuration = request.getDuration();
+
+    Collection<TimeRange> busyTimes = new HashSet<>();
+
+    // Detect relevant busy time ranges based on attendees.
+    for (Event toCheck : events) {
+      boolean attendeePresent = false;
+      for (String attendee : toCheck.getAttendees()) {
+        if (attendees.contains(attendee)) {
+          attendeePresent = true;
+        }
+      }
+      if (attendeePresent) {
+        busyTimes.add(toCheck.getWhen());
+      }
+    }
+
+    // Create an external output to hold final chopped intervals output.
+    Collection<TimeRange> actualOutput = new HashSet<>();
+    actualOutput.add(fullDay);
+
+    // Iteratively compare output intervals to busy times to chop day into final output.
+    for (TimeRange busy : busyTimes) {
+      for (TimeRange toChop : rangeOutput) {
```

Comment:
Are there any names here that could help me understand what each of these represents? I'm trying to figure out what `rangeOutput` is meant to contain, and why/what ranges are being chopped.

Reply:
Done